src/net/gosha/atproto/firehose_analysis.clj

```diff
@@ -0,0 +1,247 @@
+   [clojure.core.async         :as async]
+   [clojure.java.io            :as io]
+   [net.gosha.atproto.firehose :as firehose])
+  (:import
+   [java.time Duration Instant]))
+
+(defn calculate-message-sizes
+  [msg]
+  (if msg  ; Guard against nil messages
+    {:raw-bytes   (count (str msg))
+     :content-len (count (or (get-in msg [:record :text]) ""))}
+    {:raw-bytes   0
+     :content-len 0}))
+
+(defn format-bytes
+  [bytes]
+  (let [bytes (float (or bytes 0))]  ; Convert to float to handle ratios
+    (cond
+      (< bytes 1024)          (format "%d B" (long bytes))
+      (< bytes (* 1024 1024)) (format "%.2f KB" (double (/ bytes 1024.0)))
+      :else                   (format "%.2f MB" (double (/ bytes (* 1024.0 1024.0)))))))
+
+(defn create-window [start duration]
+  {:start-time         start
+   :end-time          (.plus start duration)
+   :message-count     0
+   :total-raw-bytes   0
+   :total-content-len 0
+   :max-msg-size      0
+   :min-msg-size      Long/MAX_VALUE
+   :types             {}})
+
+(defn create-analysis-state
+  [window-duration-seconds]
+  (let [now      (Instant/now)
+        duration (Duration/ofSeconds window-duration-seconds)]
+    {:start-time      now
+     :stop-time       nil
+     :processed-count 0
+     :current-window  (create-window now duration)
+     :windows         []
+     :window-duration duration}))
+
+(defn update-window
+  "Update window stats with a new message"
+  [window msg]
+  (if (and window msg)  ; Guard against nil window or message
+    (let [{:keys [raw-bytes content-len]} (calculate-message-sizes msg)]
+      (-> window
+          (update :message-count inc)
+          (update :total-raw-bytes + raw-bytes)
+          (update :total-content-len + content-len)
+          (update :max-msg-size max raw-bytes)
+          (update :min-msg-size min raw-bytes)
+          (update-in [:types (:type msg)] (fnil inc 0))))
+    window))
+
+(defn rotate-window!
+  [{:keys [current-window window-duration] :as state}]
+  (let [now (Instant/now)]
+    (if (.isAfter now (:end-time current-window))
+      (let [new-window (create-window now window-duration)]
+        (-> state
+            (update :windows conj current-window)
+            (assoc :current-window new-window)))
+      state)))
+
+(defn process-message!
+  [state msg]
+  (-> state
+      rotate-window!
+      (update :processed-count inc)
+      (update :current-window update-window msg)))
+
+(defn start-analysis
+  [conn & {:keys [window-duration-seconds]
+           :or   {window-duration-seconds 60}}]
+  (let [analysis-ch (async/chan 1024)
+        state       (atom (create-analysis-state window-duration-seconds))
+        mult        (async/mult (:events conn))]
+    
+    (async/tap mult analysis-ch)
+    
+    (async/go-loop []
+      (when-let [msg (async/<! analysis-ch)]
+        (swap! state process-message! msg)
+        (recur)))
+    
+    {:state       state
+     :analysis-ch analysis-ch}))
```

Comment:
No point in returning the analysis channel, the go loop is already consuming it.